be/src/exec/pipeline/nljoin/spillable_nljoin_probe_operator.cpp

```diff
@@ -0,0 +1,255 @@
+        const std::vector<ExprContext*>& conjunct_ctxs, const std::vector<SlotDescriptor*>& col_types,
+        size_t probe_column_count, const std::shared_ptr<NLJoinContext>& cross_join_context)
+        : OperatorWithDependency(factory, id, "spillable_nestloop_join_probe", plan_node_id, driver_sequence),
+          _prober(join_op, join_conjuncts, conjunct_ctxs, col_types, probe_column_count),
+          _cross_join_context(cross_join_context) {}
+
+Status SpillableNLJoinProbeOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Operator::prepare(state));
+    _accumulator.set_desired_size(state->chunk_size());
+    RETURN_IF_ERROR(_prober.prepare(state, _unique_metrics.get()));
+    _spill_factory = std::make_shared<spill::SpillerFactory>();
+    _spiller = _spill_factory->create({});
+    _cross_join_context->incr_prober();
+    return Status::OK();
+}
+
+void SpillableNLJoinProbeOperator::close(RuntimeState* state) {
+    _cross_join_context->decr_prober(state);
+    Operator::close(state);
+}
+
+bool SpillableNLJoinProbeOperator::is_ready() const {
+    auto res = _cross_join_context->is_right_finished();
+    if (res) {
+        _init_chunk_stream();
+    }
+    return _cross_join_context->is_right_finished();
+}
+
+bool SpillableNLJoinProbeOperator::is_finished() const {
+    return _is_finished || (_is_finishing && _prober.probe_finished() && _build_finished());
+}
+
+bool SpillableNLJoinProbeOperator::has_output() const {
+    return _chunk_stream && _chunk_stream->has_output();
+}
+
+bool SpillableNLJoinProbeOperator::need_input() const {
+    return _prober.probe_finished() && _build_finished();
+}
+
+Status SpillableNLJoinProbeOperator::set_finishing(RuntimeState* state) {
+    // set finishing
+    _is_finishing = true;
+    return Status::OK();
+}
+
+Status SpillableNLJoinProbeOperator::set_finished(RuntimeState* state) {
+    _is_finished = true;
+    return _cross_join_context->finish_one_left_prober(state);
+}
+
+StatusOr<ChunkPtr> SpillableNLJoinProbeOperator::pull_chunk(RuntimeState* state) {
+    TRACE_SPILL_LOG << "pull_chunk:" << _driver_sequence;
+    if (_prober.probe_finished() || _build_chunk == nullptr || _build_chunk->is_empty()) {
+        auto chunk_st = _chunk_stream->get_next(state, _executor());
+        if (chunk_st.status().is_end_of_file()) {
+            _prober.reset();
+            _set_build_finished(true);
+            RETURN_IF_ERROR(_chunk_stream->reset(state, _spiller.get()));
+            return nullptr;
+        }
+        ASSIGN_OR_RETURN(_build_chunk, std::move(chunk_st));
+
+        if (_build_chunk == nullptr || _build_chunk->is_empty()) {
+            return nullptr;
+        }
+
+        _prober.reset_probe();
+    }
+
+    ASSIGN_OR_RETURN(auto res, _prober.probe_chunk(state, _build_chunk));
+    RETURN_IF_ERROR(eval_conjuncts(_prober.conjunct_ctxs(), res.get(), nullptr));
+
+    return res;
+}
```

Comment:
why probe side need not spilling?  it is quite different from HashJoin implementation. is there a scenario like this,
only a a few rows of build side are matched for many times(because it is not filtered finally by conjuncts), other rows are matched seldom. so retain hot rows in memory and spill cold rows to disk, then NLJoin's spill implementation would resemble HashJoin, so probe operator would need spilling too.

Reply:
For cross join (NL join) each row in probe side will scan all rows in probe side. it means we should probe all partitions for build side.

You can understand that our cross join cannot use the partitioning algorithm to handle the problem. But inner-join is able to use partitioning algorithm to handle the problem. Although we do not need to spill for the probe side of the cross join, this means that we need to read the full amount of data from the build side for each chunk generated on the probe side.

This means that the amount of data we need to read for cross join.
build side spilled bytes * probe chunks.
But the amount of data to be read for a hash join.
build side spilled bytes + probe side spilled bytes